### PR TITLE
Seed all workers when using DDP

### LINKED ISSUE #7937: worker_init_fn is not set to seed dataloaders correctly when using DDP
## 🐛 Bug

When `seed_everything(workers=True)` is called, it will set the environment variable `PL_SEED_WORKERS=1`. Consequently `Trainer` will set the `worker_init_fn` for dataloaders to `pl_worker_init_function`. It seems to me that `worker_init_fn` is not set when using DDP. The reason is that [DDPPlugin.setup_environment()](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp.py#L129) eventually runs [reset_seed()](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp.py#L212), which reads the value of the `PL_GLOBAL_SEED` environment value and [calls seed_everything()](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/seed.py#L88) with the default argument `workers=False`.

## Please reproduce using the BoringModel

It's not possible to reproduce the issue in colab, since it doesn't support DDP.

### To Reproduce

Here's a simple program that demonstrates the issue:

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning import seed_everything


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    seed_everything(1234, workers=True)
    # Sets PL_SEED_WORKERS=1
    print('PL_SEED_WORKERS=' + os.environ['PL_SEED_WORKERS'])

    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
        gpus=2,
        accelerator='ddp'  # Using accelerator='dp' works
    )
    trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    # Trainer.accelerator.setup_environment() calls DDPPlugin.setup_environment(),
    # which eventually runs seed_everything(workers=False) that sets PL_SEED_WORKERS=0
    # Consequently dataloader.worker_init_fn is not set.
    print('PL_SEED_WORKERS=' + os.environ['PL_SEED_WORKERS'])


if __name__ == '__main__':
    run()

```

### Expected behavior

I would expect `pl_worker_init_function` to be called. By printing something from the function, I can see that it's called if I use `dp` accelerator, but not if I use `ddp`. I can also notice that the environment variable `PL_SEED_WORKERS` is reset to `0` during the `Trainer.fit()` call, but I would expect it to have the value `1` in the end.

I think the correct fix would be to make [reset_seed()](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/seed.py#L81) read the `PL_SEED_WORKERS` environment variable too and pass the corresponding `workers` argument to `seed_everything()`. However, I'm not familiar enough with the code to be sure that this is correct.

Preferably `pl_worker_init_function` would also display a log message that confirms that the workers are seeded correctly.

### Environment

* CUDA:
    - GPU:
        - NVIDIA Tesla V100-SXM2-16GB
        - NVIDIA Tesla V100-SXM2-16GB
    - available:         True
    - version:           11.0
* Packages:
    - numpy:             1.19.2
    - pyTorch_debug:     True
    - pyTorch_version:   1.7.0
    - pytorch-lightning: 1.4.0dev
    - tqdm:              4.51.0
* System:
    - OS:                Linux
    - architecture:
        - 64bit
        - ELF
    - processor:         x86_64
    - python:            3.8.8
    - version:           #47-Ubuntu SMP Tue May 11 15:51:42 UTC 2021

### Additional context

Recently there was discussion about [an issue with data loading](https://github.com/pytorch/pytorch/issues/5059), where the same NumPy random seed is used across different workers. This causes the workers the use the same random numbers for data transforms. A fix was quickly introduced in PyTorch Lightning that seeds the dataloaders correctly by automatically setting the  `worker_init_fn` for dataloaders.

### PR DESCRIPTION
**Pull request by senarvi**
## What does this PR do?

When seed_everything(workers=True) is called, it will set the environment variable PL_SEED_WORKERS=1. Consequently Trainer will set the worker_init_fn for dataloaders to `pl_worker_init_function`. However, this did not work with plugins that call `reset_seed()`, since that function calls `seed_everything()` with the default argument `workers=False`. This bug fix changes `seed_everything()` so that it will read the `PL_SEED_WORKERS` environment variable and pass the correct value for `workers`.

Fixes #7937

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by senarvi** [comment]
@awaelchli feel free to suggest better log messages.

File: pytorch_lightning/utilities/seed.py [hunk 1]
```diff
@@ -84,8 +84,9 @@ def reset_seed() -> None:
     If :func:`pytorch_lightning.utilities.seed.seed_everything` is unused, this function will do nothing.
     """
     seed = os.environ.get("PL_GLOBAL_SEED", None)
+    workers = os.environ.get("PL_SEED_WORKERS", False)
     if seed is not None:
-        seed_everything(int(seed))
+        seed_everything(int(seed), bool(workers))
```
Comment (carmocca) [inline pytorch_lightning/utilities/seed.py:89]: nit:

```suggestion
        seed_everything(int(seed), workers=bool(workers))
```

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -109,6 +109,7 @@ def _worker_check(self, dataloader: DataLoader, name: str) -> None:
 
     def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
         if int(os.environ.get("PL_SEED_WORKERS", 0)) and dataloader.worker_init_fn is None:
+            rank_zero_info("Setting dataloader init function to seed all workers")
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:112]: What's the benefit of this message?

File: pytorch_lightning/utilities/seed.py [hunk 3]
```diff
@@ -100,6 +101,10 @@ def pl_worker_init_function(worker_id: int, rank: Optional = None) -> None:  # p
     process_seed = torch.initial_seed()
     # back out the base seed so we can use all the bits
     base_seed = process_seed - worker_id
+    log.debug(
+        'Initializing random number generators of process %d worker %d with base seed %d',
+        global_rank, worker_id, base_seed
```
Comment (carmocca) [inline pytorch_lightning/utilities/seed.py:106]: Use f-strings

File: pytorch_lightning/trainer/data_loading.py (hunk 2 above)
Comment (senarvi) [reply pytorch_lightning/trainer/data_loading.py:112]: > What's the benefit of this message?

You can see the discussion in the [GitHub issue](https://github.com/PyTorchLightning/pytorch-lightning/issues/7937). It's impossible to know whether the workers have been seeded correctly or not, so it would be great to have some log message that confirms that the seed has been set. The best way to confirm this is to log a message from every worker of every process. We don't want to flood the log with so many info-level messages, so that's why we write debug-level messages in `pl_worker_init_function()` and just this one info-level message here. Maybe this info-level message is redundant? I think it's most important that I can turn on debug messages to check that the it's working.

File: pytorch_lightning/utilities/seed.py (hunk 3 above)
Comment (senarvi) [reply pytorch_lightning/utilities/seed.py]: Done.

File: pytorch_lightning/trainer/data_loading.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/data_loading.py:112]: > Maybe this info-level message is redundant?

I think it is and should be `debug` level. But then you already have the one inside `pl_worker_init_function` so no point in having both.

Comment (senarvi) [reply pytorch_lightning/trainer/data_loading.py]: Removed.

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGMT. Great catch !
